### PR TITLE
Move SDB into RzUtil

### PR DESCRIPTION
**Pull request by ret2libc**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

- [ ] use dist- branch to make sure source tarball is tested.

**Closing issues**

...

### CODE REVIEW

File: librz/util/sdb/src/base64.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
+/* base64 enc/dec - MIT - Copyright 2011-2016 - pancake */
```
Comment (wargio) [inline librz/util/sdb/src/base64.c]: This file can be replaced with the util equivalent

File: librz/util/sdb/src/buffer.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
+#ifndef BUFFER_H
```
Comment (wargio) [inline librz/util/sdb/src/buffer.h:4]: same here. RzBuffer ?

File: librz/util/sdb/src/json.c [hunk 3]
```diff
@@ -0,0 +1,355 @@
+/* sdb - MIT - Copyright 2012-2016 - pancake */
```
Comment (wargio) [inline librz/util/sdb/src/json.c]: i'm kinda sure the whole json code here is unused

File: librz/include/rz_userconf.h.in [hunk 4]
```diff
@@ -48,6 +48,8 @@
 #define HAVE_COPYFILE               @HAVE_COPYFILE@
 #define HAVE_COPY_FILE_RANGE        @HAVE_COPY_FILE_RANGE@
 #define HAVE_BACKTRACE              @HAVE_BACKTRACE@
+#define HAVE_GETPID                 @HAVE_GETPID@
+#define HAVE__GETPID                @HAVE__GETPID@
```
Comment (wargio) [inline librz/include/rz_userconf.h.in:53]: why sdb needs getpid ?

**Comment by XVilka** [comment]
Please remove JSON first:
- https://github.com/rizinorg/sdb/pull/27
- https://github.com/rizinorg/rizin/pull/1118

**Comment by ret2libc** [comment]
@wargio @XVilka  I won't promise I will clean everything in this PR. I think this should be the first step so that future cleaning will be much easier, but please don't expect everything here. I will just remove the obvious stuff, but things that require more changes on "rizin side " will be done in separate PRs later I think.

**Comment by XVilka** [comment]
@ret2libc removing JSON is easy, just merge my PR before and it's done. I don't see a point of importing some files to be removed immediately after, it's better to remove them beforehand.

**Comment by ret2libc** [comment]
If the pr was ready and working why wasn't it merged before? :/

**Comment by XVilka** [comment]
@ret2libc because it was simply forgotten/overlooked. It just requires fixing the test output, and it's ready: https://github.com/rizinorg/rizin/pull/1118#pullrequestreview-762547863

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -196,6 +196,9 @@ jobs:
           export PATH=${HOME}/Library/Python/3.8/bin:${HOME}/Library/Python/3.9/bin:${HOME}/.local/bin:${PATH}
           if [ "$ASAN" == "true" ]; then
             export CFLAGS="-DASAN=1 ${CFLAGS}"
+            if [ "$CC" == "clang" ]; then
+              export LD_LIBRARY_PATH=$(dirname $(clang -print-file-name=libclang_rt.asan-x86_64.so)):${LD_LIBRARY_PATH}
+            fi
```
Comment (kazarmy) [inline .github/workflows/ci.yml:201]: IIRC for Clang (based off experience from #1910):

1. `-shared-libasan` is used to indicate that the ASAN library is a shared library. Every piece of ASAN-compiled code must tell its compiler and linker of this -- if even one isn't aware, the static version of ASAN will get linked in and things will not end well (i.e. you get `Your application is linked against incompatible ASan runtimes`).
2. As per https://github.com/systemd/systemd/blob/66a0e222937e145a1722640fab413608f565cf33/docs/TESTING_WITH_SANITIZERS.md#clang, this shared library is found using `LD_LIBRARY_PATH` by the dynamic linker (not sure why you're setting `LD_LIBRARY_PATH` here).
3. If you don't use `-shared-libasan`, then you cannot use `LD_PRELOAD` and this would XX some tests.

**Review by kazarmy — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 5 above)
Comment (kazarmy) [reply .github/workflows/ci.yml:201]: > (not sure why you're setting `LD_LIBRARY_PATH` here).

Oh it's for the sdb executable; I truly forgot about this. Now I think you have to use `readelf` and examine carefully the dynamic symbols of `sdb_native` and its supporting Rizin shared libs --- all ASAN/UBSAN symbols must be UND.

File: librz/util/sdb/src/set.c [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2019 pancake <pancake@nopcode.org>
+// SPDX-License-Identifier: MIT
```
Comment (XVilka) [inline librz/util/sdb/src/set.c:2]: Is `Set` API used somewhere? I don't recall any single place, probably can be removed, both **`.c`** and **`.h`** files.

File: librz/util/time.c [hunk 7]
```diff
@@ -15,6 +15,51 @@
 #include <rz_windows.h>
 #endif
 
+#ifdef _MSC_VER
+#pragma message("gettimeofday: Windows support is ugly here")
+#include <windows.h>
+#include <time.h>
+
+RZ_API int gettimeofday(struct timeval *p, struct timezone *tz) {
```
Comment (XVilka) [inline librz/util/time.c]: Should be `rz_time_gettimeofday()` then

File: librz/include/rz_util/rz_time.h [hunk 8]
```diff
@@ -12,6 +12,17 @@
...
 
+#ifdef _MSC_VER
+struct timeval;
+
+struct timezone {
+	int tz_minuteswest; /* minutes W of Greenwich */
+	int tz_dsttime; /* type of dst correction */
+};
+
+RZ_API int gettimeofday(struct timeval *p, struct timezone *tz);
```
Comment (XVilka) [inline librz/include/rz_util/rz_time.h]: `rz_time_gettimeofday()`

File: LICENSES/CC-PDDC.txt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+
```
Comment (XVilka) [inline LICENSES/CC-PDDC.txt:1]: Is this license compatible with `LGPLv3`?

Comment (ret2libc) [reply LICENSES/CC-PDDC.txt:1]: i hope so :D there was some public domain license stuff in sdb!

Comment (ret2libc) [reply LICENSES/CC-PDDC.txt:1]: The original files just had "public domain" (also checked the original cdb.c), which I hope is the same as https://spdx.org/licenses/CC-PDDC.html . I think if it's public domain you can do whatever you want it it.

**Comment by ret2libc** [comment]
Hey @GustavoLCR could you have a look at my last commit on windows? It makes use of unicode functions.  Does it make sense? Are they alright? If you could have a look that would help a lot, thank you!

File: librz/util/sdb/src/disk.c [hunk 10]
```diff
@@ -91,11 +91,11 @@ RZ_API bool sdb_disk_finish(Sdb *s) {
...
 #if __WINDOWS__
-	LPTSTR ndump_ = rz_utf8_to_utf16(s->ndump);
-	LPTSTR dir_ = rz_utf8_to_utf16(s->dir);
+	wchar_t *ndump_ = rz_utf8_to_utf16(s->ndump);
+	wchar_t *dir_ = rz_utf8_to_utf16(s->dir);
 
-	if (MoveFileEx(ndump_, dir_, MOVEFILE_REPLACE_EXISTING)) {
-		// eprintf ("Error 0x%02x\n", GetLastError ());
+	if (!MoveFileExW(ndump_, dir_, MOVEFILE_REPLACE_EXISTING)) {
+		eprintf ("Error 0x%02x\n", GetLastError ());
```
Comment (GustavoLCR) [inline librz/util/sdb/src/disk.c]: Unicode use is correct, but its better to use `rz_sys_perror()` here

**Review by kazarmy — COMMENTED** [review]
The ASAN part lgtm

**Comment by wargio** [comment]
Something looks very broken on [S390X](https://app.travis-ci.com/github/rizinorg/rizin/jobs/576170545)

**Comment by XVilka** [comment]
@wargio it was always like that, slightly improving over time: https://github.com/rizinorg/rizin/issues/297

File: subprojects/packagefiles/xz-5.2.5/meson.build [hunk 11]
```diff
@@ -154,6 +154,23 @@ cdata.set('HAVE_MF_BT4', 1)
 cdata.set('HAVE_MF_HC3', 1)
 cdata.set('HAVE_MF_HC4', 1)
 
+has_sysconf_cpu = cc.compiles('''#include <unistd.h>
```
Comment (XVilka) [inline subprojects/packagefiles/xz-5.2.5/meson.build:157]: Could you please extract this change and send a separate PR?

File: subprojects/rzheap/rz_jemalloc/internal/bitmap.h [hunk 12]
```diff
@@ -2,83 +2,82 @@
 #ifdef JEMALLOC_H_TYPES
 
 /* Maximum bitmap bit count is 2^LG_BITMAP_MAXBITS. */
-#define	LG_BITMAP_MAXBITS	LG_RUN_MAXREGS
-#define	BITMAP_MAXBITS		(ZU(1) << LG_BITMAP_MAXBITS)
+#define LG_BITMAP_MAXBITS LG_RUN_MAXREGS
```
Comment (XVilka) [inline subprojects/rzheap/rz_jemalloc/internal/bitmap.h]: Also this change probably better as a separate PR

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/util/sdb/src/set.c (hunk 6 above)
Comment (ret2libc) [reply librz/util/sdb/src/set.c:2]: yes, for now

File: librz/include/rz_userconf.h.in (hunk 4 above)
Comment (ret2libc) [reply librz/include/rz_userconf.h.in:53]: it does for now. not exactly sure why, but we can remove it later if not strictly necessary.

File: librz/util/sdb/src/buffer.h (hunk 2 above)
Comment (ret2libc) [reply librz/util/sdb/src/buffer.h:4]: it is very different for now... if it won't be necessary anymore later we can remove it, it's not a lot of code.

File: subprojects/packagefiles/xz-5.2.5/meson.build (hunk 11 above)
Comment (ret2libc) [reply subprojects/packagefiles/xz-5.2.5/meson.build:157]: https://github.com/rizinorg/rizin/pull/2801

File: subprojects/rzheap/rz_jemalloc/internal/bitmap.h (hunk 12 above)
Comment (ret2libc) [reply subprojects/rzheap/rz_jemalloc/internal/bitmap.h]: This is actually not necessary... my editor did more than necessary here.

**Comment by ret2libc** [comment]
Closed in favour of https://github.com/rizinorg/rizin/pull/2802
